src/stream/reader/logfile.rs

```diff
@@ -0,0 +1,373 @@
+        task_id: JobTaskId,
+        instance_id: InstanceId,
+    },
+}
+
+impl LogFile {
+    pub fn open(path: &Path) -> anyhow::Result<Self> {
+        let mut file = BufReader::new(File::open(path)?);
+        LogFile::check_header(&mut file)?;
+        let start_pos = file.stream_position()?;
+        let index = LogFile::make_index(&mut file)?;
+        Ok(LogFile {
+            file,
+            index,
+            start_pos,
+        })
+    }
+
+    fn check_header(file: &mut BufReader<File>) -> anyhow::Result<()> {
+        let mut header = [0u8; 6];
+        file.read_exact(&mut header)?;
+        if &header != b"HQ:log" {
+            anyhow::bail!("Invalid file format");
+        }
+        let version = file.read_u32::<byteorder::BigEndian>()?;
+        if version != 0 {
+            anyhow::bail!("Invalid version log file version: {}", version);
+        }
+        let _ = file.read_u64::<byteorder::BigEndian>()?; // Reserved bytes
+        let _ = file.read_u64::<byteorder::BigEndian>()?; // Reserved bytes
+        Ok(())
+    }
+
+    pub fn summary(&self) -> Summary {
+        let n_opened = self
+            .index
+            .values()
+            .map(|infos| if infos.last_instance().finished { 0 } else { 1 })
+            .sum::<u64>();
+
+        let n_streams = self
+            .index
+            .values()
+            .map(|infos| infos.instances.len() as u64)
+            .sum::<u64>();
+
+        let mut stdout_size = 0u64;
+        let mut stderr_size = 0u64;
+        let mut superseded_stdout_size = 0u64;
+        let mut superseded_stderr_size = 0u64;
+
+        for task_info in self.index.values() {
+            let info = task_info.last_instance();
+            stdout_size += info.channels[0].iter().map(|c| c.size as u64).sum::<u64>();
+            stderr_size += info.channels[1].iter().map(|c| c.size as u64).sum::<u64>();
+
+            for (_, info) in task_info.superseded() {
+                superseded_stdout_size +=
+                    info.channels[0].iter().map(|c| c.size as u64).sum::<u64>();
+                superseded_stderr_size +=
+                    info.channels[1].iter().map(|c| c.size as u64).sum::<u64>();
+            }
+        }
+
+        Summary {
+            n_tasks: self.index.len() as JobTaskCount,
+            n_streams,
+            n_opened,
+            n_superseded: n_streams - self.index.len() as u64,
+            stdout_size,
+            stderr_size,
+            superseded_stderr_size,
+            superseded_stdout_size,
+        }
+    }
+
+    pub fn cat(&mut self, opts: &CatOpts) -> anyhow::Result<()> {
+        self.file.seek(SeekFrom::Start(self.start_pos))?;
+        let selected_channel_id = match opts.channel {
+            Channel::Stdout => 0,
+            Channel::Stderr => 1,
+        };
+        let mut buffer = Vec::new();
+        let mut last_pos: i64 = self.start_pos as i64;
+        let stdout = std::io::stdout();
+        let mut lock = stdout.lock();
```

Comment:
`stdout` is by default line-buffered. Maybe we want a `BufWriter` here? If there are a lot of tasks, each one with small output, and they are interleaved, it *could* be slow to print that.

Reply:
:+1: